In [1]:
import pandas as pd
import numpy as np

In [2]:
movies = pd.read_csv("../data/movies.csv")

In [3]:
ratings = pd.read_csv("../data/ratings.csv")

In [32]:
movie_ratings = ratings.merge(movies, on="movieId")

In [33]:
movie_ratings

userId  movieId  rating   timestamp  \
0              1      296     5.0  1147880044   
1              1      306     3.5  1147868817   
2              1      307     5.0  1147868828   
3              1      665     5.0  1147878820   
4              1      899     3.5  1147868510   
...          ...      ...     ...         ...   
25000090  162541    50872     4.5  1240953372   
25000091  162541    55768     2.5  1240951998   
25000092  162541    56176     2.0  1240950697   
25000093  162541    58559     4.0  1240953434   
25000094  162541    63876     5.0  1240952515   

                                                     title  \
0                                      Pulp Fiction (1994)   
1         Three Colors: Red (Trois couleurs: Rouge) (1994)   
2         Three Colors: Blue (Trois couleurs: Bleu) (1993)   
3                                       Underground (1995)   
4                               Singin' in the Rain (1952)   
...                                                    ...   
25000090                                Ratatouille (2007)   
25000091                                  Bee Movie (2007)   
25000092                    Alvin and the Chipmunks (2007)   
25000093                           Dark Knight, The (2008)   
25000094                                       Milk (2008)   

                               genres  
0         Comedy|Crime|Drama|Thriller  
1                               Drama  
2                               Drama  
3                    Comedy|Drama|War  
4              Comedy|Musical|Romance  
...                               ...  
25000090     Animation|Children|Drama  
25000091             Animation|Comedy  
25000092              Children|Comedy  
25000093      Action|Crime|Drama|IMAX  
25000094                        Drama  

[25000095 rows x 6 columns]

In [12]:
n_reviews = movie_ratings["timestamp"].value_counts().reset_index()

In [31]:
timestamp_l = 0
timestamp_r = n_reviews["timestamp"].max()
total_sum = n_reviews["count"].sum()
fraction = 0.9
while timestamp_l + 1 < timestamp_r:
    timestamp_mid = (timestamp_l + timestamp_r) // 2
    current_sum = n_reviews[n_reviews["timestamp"] < timestamp_mid]["count"].sum()
    if current_sum < (fraction * total_sum):
        timestamp_l = timestamp_mid
    else:
        timestamp_r = timestamp_mid
print(f"{fraction} of data has timestamp less than {timestamp_l}")

0.9 of data has timestamp less than 1515033908


## Train Time Margin:
timestamp < 1421944985
## Test Time Margin:
1421944985 <= timestamp < 1515033908
## Validation Time Margin:
1515033908 <= timestamp